nn/nntree.py

```diff
@@ -0,0 +1,489 @@
+            continue
+
+        if flat_is_left_nonterminal(tok):
+            new_node = Node(tok, is_terminal=False)
+            parent.add_child(new_node)
+            stack.append(parent)
+            parent = new_node
+
+            vprint(len(stack) * "    " + new_node.name)
+            continue
+
+        # Token must be a legal right nonterminal since it is not a left token
+        # A right token must be matched by its corresponding left token
+        if not flat_is_right_nonterminal(tok) or flat_matching_nonterminal(parent.name) != tok:
+            msg = "Error: Found illegal nonterminal {}, expected right nonterminal"
+            vprint(msg.format(tok))
+
+            tree = root.children[0] if root.children else None
+            return tree, ParseResult.UNOPENED_RTOKEN
+        # Empty NONTERMINALS are not allowed
+        if not parent.has_children():
+            msg = ["{}{}".format(len(stack) * "    ", tok) for tok in parse_toks[idx:]]
+            vprint("\n".join(msg))
+            vprint("Error: Tried to close empty nonterminal {}".format(tok))
+
+            tree = root.children[0] if root.children else None
+            return tree, ParseResult.EMPTY_NONTERM
+        parent = stack.pop()
+
+    if len(stack) > 1:
+        vprint("Error: Reached end of parse tokens but stack is not empty")
+        vprint([item.name for item in stack])
+        tree = root.children[0] if root.children else None
+        return tree, ParseResult.INCOMPLETE
+
+    if not root.children:
+        return None, ParseResult.FAILURE
+    if len(root.children) == 1:
+        return root.children[0], ParseResult.SUCCESS
+
+    return root.children, ParseResult.MULTI
+
+
+def tokenize_flat_tree(parse_str):
+    return parse_str.strip().split(" ")
+
+
+def parse_tree(flat_parse_str):
+    return parse_flat_tree_to_nodes(tokenize_flat_tree(flat_parse_str))
+
+
+def parse_tree_with_text(flat_tree_str, text):
+    text_toks, parse_toks = tokenize_and_merge_possible_mw_tokens(text, flat_tree_str)
+    text_toks = [tok.txt for tok in tokenizer.tokenize(text) if tok.txt]
+    return parse_flat_tree_to_nodes(parse_toks, text_toks)
+
+
+def _json_terminal_node(tok, text="placeholder", token_index=None):
+    """ first can be:
+            abfn
+            ao
+            fn
+            fs
+            gata
+            gr
+            lo
+            no
+            person
+            pfn
+            raðnr
+            sérnafn
+            so
+            so_0
+            so_1
+            so_2
+            tala
+            to
+            töl"""
+
+    subtokens = tok.split("_")
+    first = subtokens[0].split("_", 1)[0]
+
+    tail_start = 1
+    if first == "so":
+        head = subtokens[0]
+        suffix = head[-1]
+        if "_" in head and suffix in "012":
+            tail_start += int(suffix)
+    tail = subtokens[tail_start:]
+
+    if first == "no":
+        gender = [t for t in GRAMMAR.GENDERS if t in tail][:1]
```

Comment:
A much more efficient way would be to loop through the tail variants only once, collecting them into the appropriate buckets on the go